project/Jdk9.scala

```diff
@@ -11,36 +11,87 @@
  * Copyright (C) 2017-2022 Lightbend Inc. <https://www.lightbend.com>
  */
 
-import sbt._
 import sbt.Keys._
+import sbt._
 
 object Jdk9 extends AutoPlugin {
   import JdkOptions.notOnJdk8
 
+  // The version 9 is special for any Java versions >= 9
+  // and the version 11 is special for any Java versions >= 11
+  // and the version 17 is special for any Java versions >= 17
+  // and the version 21 is special for any Java versions >= 21
+  private val supportedJavaLTSVersions = List("9", "11", "17", "21")
+
   lazy val CompileJdk9 = config("CompileJdk9").extend(Compile)
 
   lazy val TestJdk9 = config("TestJdk9").extend(Test).extend(CompileJdk9)
 
-  val SCALA_SOURCE_DIRECTORY = "scala-jdk-9"
-  val SCALA_TEST_SOURCE_DIRECTORY = "scala-jdk9-only"
-  val JAVA_SOURCE_DIRECTORY = "java-jdk-9"
-  val JAVA_TEST_SOURCE_DIRECTORY = "java-jdk9-only"
+  lazy val ScalaSourceDirectories: Seq[String] = getAdditionalSourceDirectoryNames("scala", isTest = false)
+  lazy val ScalaTestSourceDirectories: Seq[String] = getAdditionalSourceDirectoryNames("scala", isTest = true)
+
+  lazy val JavaSourceDirectories: Seq[String] = getAdditionalSourceDirectoryNames("java", isTest = false)
+  lazy val JavaTestSourceDirectories: Seq[String] = getAdditionalSourceDirectoryNames("java", isTest = true)
+
+  private lazy val additionalScalaSourceDirectories = Def.setting {
+    getAdditionalSourceDirectories(
+      (Compile / sourceDirectory).value,
+      ScalaSourceDirectories)
+  }
+
+  private lazy val additionalJavaSourceDirectories = Def.setting {
+    getAdditionalSourceDirectories(
+      (Compile / sourceDirectory).value,
+      JavaSourceDirectories)
+  }
+
+  lazy val additionalSourceDirectories = Def.setting {
+    additionalScalaSourceDirectories.value ++ additionalJavaSourceDirectories.value
+  }
+
+  private lazy val additionalScalaTestSourceDirectories = Def.setting {
+    getAdditionalSourceDirectories(
+      (Test / sourceDirectory).value,
+      ScalaTestSourceDirectories)
+  }
+
+  private lazy val additionalJavaTestSourceDirectories = Def.setting {
+    getAdditionalSourceDirectories(
+      (Test / sourceDirectory).value,
+      JavaTestSourceDirectories)
+  }
+
+  lazy val additionalTestSourceDirectories = Def.setting {
+    additionalScalaTestSourceDirectories.value ++ additionalJavaTestSourceDirectories.value
+  }
+
+  private def getAdditionalSourceDirectoryNames(language: String, isTest: Boolean): Seq[String] = {
+    for {
+      version <- supportedJavaLTSVersions if version.toInt <= JdkOptions.JavaVersion.majorVersion
+    } yield {
+      if (isTest) {
+        s"$language-jdk${version}-only"
```

Comment:
I notice you are using string interpolation, why not use `suffix` argument rather than a boolean, we could gain beautiful code.